presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcTypeToHiveTypeTranslator.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.orc;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.orc.metadata.ColumnMetadata;
+import io.prestosql.orc.metadata.OrcType;
+import io.prestosql.plugin.hive.HiveType;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.prestosql.plugin.hive.HiveType.HIVE_BINARY;
+import static io.prestosql.plugin.hive.HiveType.HIVE_BOOLEAN;
+import static io.prestosql.plugin.hive.HiveType.HIVE_BYTE;
+import static io.prestosql.plugin.hive.HiveType.HIVE_DATE;
+import static io.prestosql.plugin.hive.HiveType.HIVE_DOUBLE;
+import static io.prestosql.plugin.hive.HiveType.HIVE_FLOAT;
+import static io.prestosql.plugin.hive.HiveType.HIVE_INT;
+import static io.prestosql.plugin.hive.HiveType.HIVE_LONG;
+import static io.prestosql.plugin.hive.HiveType.HIVE_SHORT;
+import static io.prestosql.plugin.hive.HiveType.HIVE_STRING;
+import static io.prestosql.plugin.hive.HiveType.HIVE_TIMESTAMP;
+import static io.prestosql.plugin.hive.HiveType.toHiveType;
+import static org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory.getListTypeInfo;
+import static org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory.getMapTypeInfo;
+import static org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory.getStructTypeInfo;
+
+public class OrcTypeToHiveTypeTranslator
+{
+    private OrcTypeToHiveTypeTranslator() {}
+
+    public static HiveType fromOrcTypeToHiveType(OrcType orcType, ColumnMetadata<OrcType> columnMetadata)
```

Comment:
I'm also uncomfortable with the multiple type conversions.  By far the messiest part of supporting Hive ACID delete and insert is conversions among the three type systems: Presto Types, HiveTypes and OrcTypes.

However, as pointed out by @electrum, we need this specific translation to ensure to ensure we can read Hive tables even if some other agent inserted data that didn't have all the columns we expect.  When we read, the metadata must come only from the file itself, and there it is stored as OrcTypes.

Reply:
Sure, it makes sense to replicate the _data_ ORC file's schema to _delete delta_ ORC file's schema.
The code should do that as directly as possible -- so we should strive not to go thru Hive types.

(the mapping may be not 1-1 and so going thru hive types may erase some information; or, it may be some funky ORC type we do not understand at all yet)